### PR TITLE
Support returning with structures under torch.return_types

### PR DESCRIPTION
**Pull request by mavenlin**

### CODE REVIEW

File: torch2jax/__init__.py [hunk 1]
```diff
@@ -415,6 +421,23 @@ def flatten(input, start_dim=0, end_dim=-1):
   return jnp.reshape(_v(input), input.shape[:start_dim] + (-1,))
 
 
+@implements(torch.max, out_kwarg=True, Torchish_member=True)
+def max(input, *args):
```
Comment (mavenlin) [inline torch2jax/__init__.py]: Discussion: 
how can we handle the function overloading? the current implementation works okay if all args are pos args. But would fail if we call `max(a, dim=1)` or `max(a, other=torch.tensor(...))`.

Comment (samuela) [reply torch2jax/__init__.py]: Yeah, this is a real PITA... the best I've come up with thus far is to accept `*args` and `**kwargs` and do some manual parsing from there. see eg https://github.com/samuela/torch2jax/blob/cb1c4cab9dec465c3d0acc57b40e607d21ae391a/torch2jax/__init__.py#L355-L376. it's not pretty, and improvements are welcome

File: torch2jax/__init__.py [hunk 2]
```diff
@@ -618,7 +641,10 @@ def scatter_add(input, dim, index, src):
 
 @implements(torch.sort, out_kwarg=True, Torchish_member=True)
 def sort(input, dim=-1, descending=False, stable=False):
-  return jnp.sort(_v(input), axis=dim, stable=stable, descending=descending)
+  input = _v(input)
+  sorted_indices = jnp.argsort(input, axis=dim, stable=stable, descending=descending)
+  sorted_values = jnp.take_along_axis(input, sorted_indices, axis=dim)
+  return torch.return_types.sort([sorted_values, sorted_indices])
```
Comment (samuela) [inline torch2jax/__init__.py:655]: this changes the return type, so the previous impl was wrong? do we not have tests for sort?

File: torch2jax/__init__.py [hunk 3]
```diff
@@ -270,14 +270,20 @@ def implements(torch_function, Torchishify_output=True, out_kwarg=False, Torchis
...
     if out_kwarg:
 
+      def assign(a, b):
+        a.value = b
+
       def func1(*args, out=None, **kwargs):
         if out is not None:
-          out.value = func(*args, **kwargs)
+          ret = func(*args, **kwargs)
+          torch_tree_map(assign, out, ret)
```
Comment (samuela) [inline torch2jax/__init__.py:281]: nit: let's add a comment here giving an example of a function that requires a non-array `out` value

File: torch2jax/__init__.py [hunk 4]
```diff
@@ -270,14 +270,20 @@ def implements(torch_function, Torchishify_output=True, out_kwarg=False, Torchis
...
           return out
         else:
-          return Torchish(func(*args, **kwargs))
+          return torch_tree_map(Torchish, func(*args, **kwargs))
     elif Torchishify_output:
-      func1 = lambda *args, **kwargs: Torchish(func(*args, **kwargs))
+      # although func is jax code, but it is used to mock the torch function, therefore the returning container (if any) should be torch types
+      # e.g. torch.return_types.max, therefore, we use torch_tree_map here instead of jax.tree.map
```
Comment (samuela) [inline torch2jax/__init__.py:287]: although to be fair we're registering all these types as jax pytrees anyways, so could we use jax.tree.map after all?

File: torch2jax/__init__.py (hunk 2 above)
Comment (mavenlin) [reply torch2jax/__init__.py:655]: I don't find an existing test

File: torch2jax/__init__.py (hunk 4 above)
Comment (mavenlin) [reply torch2jax/__init__.py:287]: Registration happens only when casting back to Jax. So it could be unregistered when we call this line

File: torch2jax/__init__.py (hunk 2 above)
Comment (samuela) [reply torch2jax/__init__.py:655]: ok, let's create one so that we don't mess this up again in the future

File: torch2jax/__init__.py (hunk 4 above)
Comment (samuela) [reply torch2jax/__init__.py:287]: ah ok that makes sense. let's add that detail to the comment here for future reference

File: torch2jax/__init__.py (hunk 1 above)
Comment (mavenlin) [reply torch2jax/__init__.py]: Done

**Comment by mavenlin** [comment]
@samuela I've resolved the problems, could you take another look? Thanks!

File: tests/utils.py [hunk 5]
```diff
@@ -33,13 +34,13 @@ def out_kwarg_test(f, args, kwargs={}, **assert_kwargs):
...
     out1 = f(*torch_args, **torch_kwargs)
-    out2 = torch.zeros_like(out1)
+    out2 = torch_tree_map(torch.zeros_like, out1)
     out3 = f(*torch_args, out=out2, **torch_kwargs)
     return out1, out2, out3
 
   jax_function = t2j(torch_function)
   jax_out1, jax_out2, jax_out3 = jax_function(*args)
-  assert jax_out2 is jax_out3
+  assert all(jax.tree.leaves(jax.tree.map(lambda x, y: x is y, jax_out2, jax_out3)))
```
Comment (samuela) [inline tests/utils.py:43]: is it actually necessary to change this? i would expect `assert jax_out2 is jax_out3` to work as well

**Review by samuela — COMMENTED** [review]
curious about this, but o/w lgtm

File: tests/utils.py (hunk 5 above)
Comment (mavenlin) [reply tests/utils.py:43]: Because when converting to Jax the out2 and out3 gets converted separately, but the array stored inside is the same copy.

**Comment by mavenlin** [comment]
Hi @samuela, do you plan to merge this soon? As I have some more PRs that are dependent on this one.

File: tests/utils.py (hunk 5 above)
Comment (samuela) [reply tests/utils.py:43]: ok i dug into this a bit more: although `@implements` returns the `out` kwarg when it is non-None, `t2j_function` calls `_tree_coerce` on the result and this calls `jax.tree.map` which changes object identity.

**Review by samuela — APPROVED** [review]
LGTM, thanks @mavenlin !
